packages/graph/hash_graph/lib/graph/src/api/web/data_type.rs

```diff
@@ -0,0 +1,145 @@
+//! Web routes for Data Types CRU operations.
+
+use axum::{
+    extract::Path,
+    http::StatusCode,
+    response::IntoResponse,
+    routing::{get, post},
+    Extension, Json, Router,
+};
+use serde::{Deserialize, Serialize};
+use utoipa::{Component, OpenApi};
+use uuid::Uuid;
+
+use super::{
+    api_resource::RoutedResource,
+    error::{modify_report_to_status_code, query_report_to_status_code},
+};
+use crate::{
+    datastore::Datastore,
+    types::{schema::DataType, AccountId, BaseId, Qualified, QualifiedDataType, VersionId},
+};
+
+#[derive(OpenApi)]
+#[openapi(
+    handlers(
+        create_data_type,
+        get_data_type,
+        // get_data_type_many,
+        update_data_type
+    ),
+    components(CreateDataTypeRequest, UpdateDataTypeRequest, AccountId, BaseId, QualifiedDataType),
+    tags(
+        (name = "DataType", description = "Data Type management API")
+    )
+)]
+pub struct DataTypeResource;
+
+impl RoutedResource for DataTypeResource {
+    /// Create routes for interacting with data types.
+    fn routes<D: Datastore>() -> Router {
+        // TODO: The URL format here is preliminary and will have to change.
+        Router::new().nest(
+            "/data-type",
+            Router::new()
+                .route("/", post(create_data_type::<D>).put(update_data_type::<D>))
+                // .route("/query", get(get_data_type_many))
+                .route("/:version_id", get(get_data_type::<D>)),
+        )
+    }
+}
+
+#[derive(Serialize, Deserialize, Component)]
+struct CreateDataTypeRequest {
+    #[component(value_type = Any)]
+    schema: DataType,
+    account_id: AccountId,
+}
+
+#[utoipa::path(
+    post,
+    path = "/data-type",
+    request_body = CreateDataTypeRequest,
+    tag = "DataType",
+    responses(
+      (status = 201, content_type = "application/json", description = "Data type created successfully", body = QualifiedDataType),
+      (status = 422, content_type = "text/plain", description = "Provided request body is invalid"),
+
+      (status = 409, description = "Unable to create data type in the datastore as the base data type ID already exists"),
+      (status = 500, description = "Datastore error occurred"),
+    ),
+    request_body = CreateDataTypeRequest,
+)]
+async fn create_data_type<D: Datastore>(
+    Json(body): Json<CreateDataTypeRequest>,
+    Extension(datastore): Extension<D>,
```

Comment:
This syntax is confusing 👀

Reply:
It is, yeah. All the routing extensions in Axum are defined by 1-tuple structs. Would you prefer to destructure the `extension` in the body, or accessing by `.0`? 
I don't mind the syntax too much personally :smile:

Reply:
Oh, tuple structs 👁️👁️ 
I'd prefer not to do it like this. Either other way seems fine

Reply:
Addressed in https://github.com/hashintel/hash/pull/750/commits/35a65f91e7b982c6fe7e6b3f77178486138630ea